File: servers/typescript/src/handler_connection.ts
import type { StarfishFrame } from "./types.js";
import type { Client } from "./client.js";
import type { StarfishServer } from "./starfish_server.js";
import type { ResumeEntry } from "./resume.js";
import { createErrorFrame, ERR_PAYLOAD_TOO_LARGE, ERR_PROTOCOL_UNSUPPORTED_VERSION } from "./errors.js";
import { MAX_CLIENT_META_SIZE } from "./limits.js";

const SUPPORTED_VERSIONS = [1];

type HelloPayload = {
  versions?: number[];
  client?: { name?: string; role?: string; meta?: unknown };
  capabilities?: { rtc?: boolean };
  resumeToken?: string;
  auth?: { type?: string };
};

type WelcomePayload = {
  status: string;
  version: number;
  clientId: string;
  resumed?: boolean;
  resumeToken: string;
  resumeTimeout: number;
  serverTime: number;
  heartbeatInterval: number;
  sessionRequired: boolean;
  sessions?: string[];
  pools?: string[];
  rtc?: { iceServers: Array<{ urls: string }> };
};

function negotiateVersion(clientVersions: number[]): number | undefined {
  // A client that offers no versions (e.g. a resume hello) is accepted at the
  // default supported version for backwards compatibility.
  if (clientVersions.length === 0) return SUPPORTED_VERSIONS[0];
  for (const v of clientVersions) {
    if (SUPPORTED_VERSIONS.includes(v)) return v;
  }
  return undefined;
}

export function handleClientHello(
  hub: StarfishServer,
  client: Client,
  frame: StarfishFrame,
): void {
  const payload = (frame.payload ?? {}) as HelloPayload;

  const clientVersions = payload.versions ?? [];
  const negotiated = negotiateVersion(clientVersions);
  if (negotiated === undefined) {
    client.sendFrame(
      createErrorFrame(hub.idGen, frame.header.id, ERR_PROTOCOL_UNSUPPORTED_VERSION, "client", "welcome"),
    );
    return;
  }

  if (payload.resumeToken) {
    const resumed = handleResume(hub, client, frame, payload.resumeToken, negotiated);
    if (resumed) return;
  }

  handleFreshHello(hub, client, frame, payload, negotiated);
}

function handleResume(
  hub: StarfishServer,
  client: Client,
  frame: StarfishFrame,
  token: string,
  version: number,
): boolean {
  const entry = hub.resumes.restore(token);
  if (!entry) {
    return false;
  }

  restoreClient(client, entry);
  hub.registerClient(client);
  rejoinSessions(hub, client, entry);

  const newToken = hub.idGen.resumeToken();
  hub.resumes.registerToken(client, newToken);

  sendWelcome(hub, client, frame, version, {
    clientId: client.id,
    resumed: true,
    resumeToken: newToken,
    sessions: [...client.sessions],
    pools: [...client.pools],
  });

  return true;
}

function handleFreshHello(
  hub: StarfishServer,
  client: Client,
  frame: StarfishFrame,
  payload: HelloPayload,
  version: number,
): void {
  const clientId = hub.idGen.clientId();
  const resumeToken = hub.idGen.resumeToken();

  client.id = clientId;
  if (payload.client) {
    if (payload.client.meta !== undefined) {
      const metaSize = JSON.stringify(payload.client.meta).length;
      if (metaSize > MAX_CLIENT_META_SIZE) {
        client.sendFrame(createErrorFrame(hub.idGen, frame.header.id, ERR_PAYLOAD_TOO_LARGE, "client", "welcome"));
        return;
      }
    }
    client.name = payload.client.name ?? "";
    client.role = payload.client.role ?? "";
    client.meta = payload.client.meta;
  }
  if (payload.capabilities) {
    client.rtcCapable = payload.capabilities.rtc === true;
  }
  client.authenticated = true;
  client.lastActivity = Date.now();

  hub.registerClient(client);
  hub.resumes.registerToken(client, resumeToken);

  sendWelcome(hub, client, frame, version, {
    clientId,
    resumeToken,
    sessionRequired: true,
  });
}

function restoreClient(client: Client, entry: ResumeEntry): void {
  Object.assign(client, {
    id: entry.clientId, name: entry.name, role: entry.role, meta: entry.meta,
    rtcCapable: entry.rtcCapable, sessions: entry.sessions, topics: entry.topics,
    authenticated: true, lastActivity: Date.now(),
  });
}

function rejoinSessions(hub: StarfishServer, client: Client, entry: ResumeEntry): void {
  for (const sessName of client.sessions) {
    const sess = hub.getSession(sessName);
    if (!sess) continue;
    sess.addClient(client);
    const topicSet = client.topics.get(sessName);
    if (topicSet) {
      for (const topic of topicSet) sess.subscribe(topic, client);
    }
    const presenceData = entry.presence.get(sessName);
    if (presenceData !== undefined) sess.setPresence(client.id, presenceData);
  }
}

function sendWelcome(
  hub: StarfishServer,
  client: Client,
  frame: StarfishFrame,
  version: number,
  extra: Partial<WelcomePayload>,
): void {
  const now = Date.now();
  const welcomePayload: WelcomePayload = {
    status: "ok",
    version,
    clientId: client.id,
    resumeToken: "",
    resumeTimeout: hub.config.resumeTimeoutMs,
    serverTime: now,
    heartbeatInterval: hub.config.heartbeatIntervalMs,
    sessionRequired: false,
    ...extra,
  };

  if (hub.config.iceServers.length > 0) {
    welcomePayload.rtc = { iceServers: hub.config.iceServers };
  }

  client.sendFrame({
    header: {
      v: 1,
      id: hub.idGen.messageId(),
      resource: "client",
      method: "welcome",
      kind: "response",
      ts: now,
      replyTo: frame.header.id,
    },
    payload: welcomePayload as unknown as Record<string, unknown>,
  });
}

File: servers/python/src/starfish_server/handler_connection.py
from __future__ import annotations

import json
import time
from typing import TYPE_CHECKING, Any

from .client import Client
from .errors import ERR_PAYLOAD_TOO_LARGE, ERR_PROTOCOL_UNSUPPORTED_VERSION, create_error_frame
from .limits import MAX_CLIENT_META_SIZE
from .types import StarfishFrame

if TYPE_CHECKING:
    from .resume import ResumeEntry
    from .server import StarfishServer

SUPPORTED_VERSIONS = [1]


def handle_client_hello(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    payload = frame.get("payload") or {}
    header = frame.get("header", {})

    client_versions = payload.get("versions", [])
    negotiated = _negotiate_version(client_versions)
    if negotiated is None:
        client.send_frame(
            create_error_frame(hub.id_gen, header.get("id", ""), ERR_PROTOCOL_UNSUPPORTED_VERSION, "client", "welcome")
        )
        return

    resume_token = payload.get("resumeToken")
    if resume_token:
        if _handle_resume(hub, client, frame, resume_token, negotiated):
            return

    _handle_fresh_hello(hub, client, frame, payload, negotiated)


def _negotiate_version(client_versions: list[int]) -> int | None:
    # A client that offers no versions (e.g. a resume hello) is accepted at the
    # default supported version for backwards compatibility.
    if not client_versions:
        return SUPPORTED_VERSIONS[0]
    for v in client_versions:
        if v in SUPPORTED_VERSIONS:
            return v
    return None


def _handle_resume(
    hub: StarfishServer,
    client: Client,
    frame: StarfishFrame,
    token: str,
    version: int,
) -> bool:
    entry = hub.resumes.restore(token)
    if entry is None:
        return False

    _restore_client(client, entry)
    hub.register_client(client)
    _rejoin_sessions(hub, client, entry)

    new_token = hub.id_gen.resume_token()
    hub.resumes.register_token(client, new_token)

    _send_welcome(hub, client, frame, version, {
        "clientId": client.id,
        "resumed": True,
        "resumeToken": new_token,
        "sessions": list(client.sessions),
        "pools": list(client.pools),
    })
    return True


def _handle_fresh_hello(
    hub: StarfishServer,
    client: Client,
    frame: StarfishFrame,
    payload: dict[str, Any],
    version: int,
) -> None:
    header = frame.get("header", {})
    client_id = hub.id_gen.client_id()
    resume_token = hub.id_gen.resume_token()

    client.id = client_id
    client_info = payload.get("client", {})
    if client_info:
        meta = client_info.get("meta")
        if meta is not None:
            meta_size = len(json.dumps(meta))
            if meta_size > MAX_CLIENT_META_SIZE:
                client.send_frame(
                    create_error_frame(hub.id_gen, header.get("id", ""), ERR_PAYLOAD_TOO_LARGE, "client", "welcome")
                )
                return
        client.name = client_info.get("name", "")
        client.role = client_info.get("role", "")
        client.meta = meta

    capabilities = payload.get("capabilities", {})
    if capabilities:
        client.rtc_capable = capabilities.get("rtc", False) is True

    client.authenticated = True
    client.last_activity = _now_ms()

    hub.register_client(client)
    hub.resumes.register_token(client, resume_token)

    _send_welcome(hub, client, frame, version, {
        "clientId": client_id,
        "resumeToken": resume_token,
        "sessionRequired": True,
    })


def _restore_client(client: Client, entry: ResumeEntry) -> None:
    client.id = entry.client_id
    client.name = entry.name
    client.role = entry.role
    client.meta = entry.meta
    client.rtc_capable = entry.rtc_capable
    client.sessions = set(entry.sessions)
    client.topics = {k: set(v) for k, v in entry.topics.items()}
    client.authenticated = True
    client.last_activity = _now_ms()


def _rejoin_sessions(hub: StarfishServer, client: Client, entry: ResumeEntry) -> None:
    for sess_name in client.sessions:
        sess = hub.get_session(sess_name)
        if sess is None:
            continue
        sess.add_client(client)
        topic_set = client.topics.get(sess_name)
        if topic_set:
            for topic in topic_set:
                sess.subscribe(topic, client)
        presence_data = entry.presence.get(sess_name)
        if presence_data is not None:
            sess.set_presence(client.id, presence_data)


def _send_welcome(
    hub: StarfishServer,
    client: Client,
    frame: StarfishFrame,
    version: int,
    extra: dict[str, Any],
) -> None:
    now = _now_ms()
    payload: dict[str, Any] = {
        "status": "ok",
        "version": version,
        "clientId": client.id,
        "resumeToken": "",
        "resumeTimeout": hub.config.resume_timeout_ms,
        "serverTime": now,
        "heartbeatInterval": hub.config.heartbeat_interval_ms,
        "sessionRequired": False,
        **extra,
    }

    if hub.config.ice_servers:
        payload["rtc"] = {"iceServers": [{"urls": s.urls} for s in hub.config.ice_servers]}

    client.send_frame({
        "header": {
            "v": 1,
            "id": hub.id_gen.message_id(),
            "resource": "client",
            "method": "welcome",
            "kind": "response",
            "ts": now,
            "replyTo": frame.get("header", {}).get("id", ""),
        },
        "payload": payload,
    })


def _now_ms() -> int:
    return int(time.time() * 1000)
